From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 28 Jun 2020 19:08:41 -0400
Subject: [PATCH] Improve Legacy Component serialization size

Don't constantly send format: false for all formatting options when parent already
has it false

diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftChatMessage.java b/src/main/java/org/bukkit/craftbukkit/util/CraftChatMessage.java
index b27af66795d902a2e95d692fa0ff18eccbef8a75..b89660244ffad484e3fbf69ccb9cdc1bc178d2ad 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftChatMessage.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftChatMessage.java
@@ -46,6 +46,7 @@ public final class CraftChatMessage {
         // Separate pattern with no group 3, new lines are part of previous string
         private static final Pattern INCREMENTAL_PATTERN_KEEP_NEWLINES = Pattern.compile("(" + String.valueOf(org.bukkit.ChatColor.COLOR_CHAR) + "[0-9a-fk-orx])|((?:(?:https?):\\/\\/)?(?:[-\\w_\\.]{2,}\\.[a-z]{2,4}.*?(?=[\\.\\?!,;:]?(?:[" + String.valueOf(org.bukkit.ChatColor.COLOR_CHAR) + " ]|$))))", Pattern.CASE_INSENSITIVE);
         // ChatColor.b does not explicitly reset, its more of empty
+        private static final Style EMPTY = Style.EMPTY.withItalic(false); // Paper - OBFHELPER
         private static final Style RESET = Style.EMPTY.withBold(false).withItalic(false).setUnderline(false).setStrikethrough(false).setRandom(false);
 
         private final List<Component> list = new ArrayList<Component>();
@@ -67,6 +68,7 @@ public final class CraftChatMessage {
             Matcher matcher = (keepNewlines ? INCREMENTAL_PATTERN_KEEP_NEWLINES : INCREMENTAL_PATTERN).matcher(message);
             String match = null;
             boolean needsAdd = false;
+            boolean hasReset = false; // Paper
             while (matcher.find()) {
                 int groupId = 0;
                 while ((match = matcher.group(++groupId)) == null) {
@@ -112,7 +114,26 @@ public final class CraftChatMessage {
                             throw new AssertionError("Unexpected message format");
                         }
                     } else { // Color resets formatting
-                        modifier = RESET.withColor(format);
+                        // Paper start - improve legacy formatting
+                        Style previous = modifier;
+                        modifier = (!hasReset ? RESET : EMPTY).withColor(format);
+                        hasReset = true;
+                        if (previous.isBold()) {
+                            modifier = modifier.withBold(false);
+                        }
+                        if (previous.isItalic()) {
+                            modifier = modifier.withItalic(false);
+                        }
+                        if (previous.isObfuscated()) {
+                            modifier = modifier.setRandom(false);
+                        }
+                        if (previous.isStrikethrough()) {
+                            modifier = modifier.setStrikethrough(false);
+                        }
+                        if (previous.isUnderlined()) {
+                            modifier = modifier.setUnderline(false);
+                        }
+                        // Paper end
                     }
                     needsAdd = true;
                     break;
